### ISSUE TITLE
Should the router handle `params` per handler?

### ISSUE THREAD

**Issue by yusukebe** [honojs/hono#1141]
This issue pertains to the router. Currently, the result object is composed of `handlers` and `params`:

```ts
interface Result<T> {
  handlers: T[]
  params: Record<string, string>
}
```

However, it might be that each `handler` should have its own `params`, rather than all handlers sharing one `params`.

For instance, in the current pattern, the result is not as expected. (This example uses TrieRouter, because RegExpRouter has another issue #1140)

```ts
app.get('/webhooks/github', async (c, next) => {
  await next()
})

app.get('/webhooks/:service', (c) => {
  return c.text(`service: ${c.req.param('service')}`)
})
```

```
GET /webhooks/github => service: undefined
```

I believe it should return `service: github`. To make this work correctly, we might need a structure like the one below, where `params` are handled per handler.

```ts
interface Result<T> {
  routes: Route<T>[]
}

interface Route<T> {
  handler: T
  params: Record<string, string>
}
```

If we decide to address these issues, it will require significant effort. But, it might be worth doing.

**Comment by yusukebe** [comment]
@usualoma 

What are your thoughts on this spec?

**Comment by yusukebe** [comment]
I've reconsidered this.

I believe we don't need to make the above changes. It's okay to share the params among all matched handlers. I recall this was discussed before.

So, what we need to do is fix the issue where `GET /webhooks/github` returns `service: undefined`. I've created a PR to fix this for TrieRouter #1145.

**Comment by usualoma** [comment]
Hi @yusukebe!

It is one idea worth considering, but I agree that we don't have to do it.
